cdap-security/src/main/java/co/cask/cdap/security/auth/context/AuthenticationContextModules.java

```diff
@@ -44,13 +53,26 @@ protected void configure() {
 
   /**
    * An {@link AuthenticationContext} for use in program containers. The authentication details in this context are
-   * determined based on the {@link UserGroupInformation} of the user running the program.
+   * determined based on the {@link UserGroupInformation} of the user running the program. Additionally,
+   * if an principal was provided then that information is also included in the {@link Principal}.
    */
-  public AbstractModule getProgramContainerModule() {
+  public AbstractModule getProgramContainerModule(@Nullable final String principal) {
     return new AbstractModule() {
       @Override
       protected void configure() {
-        bind(AuthenticationContext.class).to(ProgramContainerAuthenticationContext.class);
+        bind(AuthenticationContext.class).toInstance(new AuthenticationContext() {
+          @Override
+          public Principal getPrincipal() {
+            try {
+              return principal == null ?
+                new Principal(UserGroupInformation.getCurrentUser().getShortUserName(), Principal.PrincipalType.USER) :
```

Comment:
Strictly speaking, shouldn't get the current user from here, as there is no guarantee who that will be. I think the UGI should be passed-in or determined when the `getProgramContainerModule` method is called:

```java
// BTW, that's how you avoid caller to pass in null
Module getProgramContainerModule() {
  String userName = UserGroupInformation.getCurrentUser().getShortUserName();
  return new ProgramContainerModule(new Principal(userName, Principal.PrincipalType.USER));
}

Module getProgramContainerModule(String principal) {
  String userName = UserGroupInformation.getCurrentUser().getShortUserName();
  return new ProgramContainerModule(new Principal(userName, Principal.PrincipalType.USER, principal));
}
```